svcb.go

```diff
@@ -0,0 +1,795 @@
+		if len(e) != net.IPv6len || e.To4() != nil {
+			return nil, errors.New("dns: not IPv6")
+		}
+		b = append(b, e...)
+	}
+	return b, nil
+}
+
+func (s *SVCBIPv6Hint) unpack(b []byte) error {
+	if len(b) == 0 || len(b)%16 != 0 {
+		return errors.New("dns: bad array of IPv6 addresses")
+	}
+	i := 0
+	x := make([]net.IP, 0, len(b)/16)
+	for i < len(b) {
+		x = append(x, net.IP(b[i:i+16]))
+		i += 16
+	}
+	s.Hint = x
+	return nil
+}
+
+// String returns "<nil>" if an invalid IPv6 address was encountered.
+// TODO DOC Do I need full definition for doc?
+func (s *SVCBIPv6Hint) String() string {
+	var str strings.Builder
+	str.Grow(40 * len(s.Hint))
+	for _, e := range s.Hint {
+		if e.To4() != nil {
+			return "<nil>"
+		}
+		str.WriteByte(',')
+		str.WriteString(e.String())
+	}
+	return str.String()[1:]
+}
+
+func (s *SVCBIPv6Hint) parse(b string) error {
+	if strings.ContainsRune(b, '.') {
+		return errors.New("dns: not IPv6")
+	}
+	str := strings.Split(b, ",")
+	dst := make([]net.IP, 0, len(str))
+	for _, e := range str {
+		ip := net.ParseIP(e)
+		if ip == nil {
+			return errors.New("dns: bad IP")
+		}
+		dst = append(dst, ip)
+	}
+	s.Hint = dst
+	return nil
+}
+
+func (s *SVCBIPv6Hint) copy() SVCBKeyValue {
+	return &SVCBIPv6Hint{
+		append(make([]net.IP, 0, len(s.Hint)), s.Hint...),
+	}
+}
+
+// SVCBLocal pair is intended for experimental/private use.
+// The key is recommended to be in the range
+// [SVCB_PRIVATE_LOWER, SVCB_PRIVATE_UPPER].
+// Basic use pattern for creating a keyNNNNN option:
+//
+//	o := new(dns.HTTPS)
+//	o.Hdr.Name = "."
+//	o.Hdr.Rrtype = dns.HTTPS
+//	e := new(dns.SVCBLocal)
+//	e.KeyCode = 65400
+//	e.Data = []byte("abc")
+//	o.Value = append(o.Value, e)
+type SVCBLocal struct {
+	KeyCode uint16 // Never 65535 or any assigned keys
+	Data    []byte // All byte sequences are allowed
+	// For the string representation, See draft-ietf-dnsop-svcb-https
+	// (TODO RFC XXXX)
+	// "2.1.1.  Presentation format for SVCBFieldValue key=value pairs"
+	// for a full list of allowed characters. Otherwise escape codes
+	// e.g. \000 for NUL and \127 for DEL are used.
+}
+
+func (s *SVCBLocal) Key() uint16           { return s.KeyCode }
+func (s *SVCBLocal) pack() ([]byte, error) { return s.Data, nil }
+func (s *SVCBLocal) unpack(b []byte) error { s.Data = b; return nil }
+func (s *SVCBLocal) len() uint16           { return uint16(len(s.Data)) }
+
+// String escapes whitespaces too, which is not required when
+// the result would be enclosed in double quotes. TODO Is this doc fine?
+// do i need definition
+func (s *SVCBLocal) String() string {
+	var str strings.Builder
+	str.Grow(4 * len(s.Data))
+	for _, e := range s.Data {
+		if (0x1f < e && e < 0x7f) || e == 0x09 {
+			switch e {
+			case '"':
+				fallthrough
+			case ';':
+				fallthrough
+			// As promised, optionally handle space
+			case ' ':
+				fallthrough
+			// Tab
+			case 0x09:
+				fallthrough
+			case '\\':
+				str.WriteByte('\\')
+				fallthrough
```

Comment:
Drop the `fallthrough` and just use `str.WriteByte(e)`.

Reply:
I can't drop the `fallthrough` but now it looks like this:
```go
			switch e {
			case '"', ';', ' ', '\\', 0x09:
				str.WriteByte('\\')
				fallthrough
			default:
				str.WriteByte(e)
			}
```

Reply:
What I'm saying is replace this with:
```go
case ...:
	str.WriteByte('\\')
	str.WriteByte(e)
default:
	str.WriteByte(e)
```